pkg/ccl/importccl/read_import_mysql.go

```diff
@@ -0,0 +1,144 @@
+// Copyright 2018 The Cockroach Authors.
+//
+// Licensed as a CockroachDB Enterprise file under the Cockroach Community
+// License (the "License"); you may not use this file except in compliance with
+// the License. You may obtain a copy of the License at
+//
+//     https://github.com/cockroachdb/cockroach/blob/master/licenses/CCL.txt
+
+package importccl
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/pkg/errors"
+	mysql "github.com/xwb1989/sqlparser"
+
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/types"
+	"github.com/cockroachdb/cockroach/pkg/sql/sessiondata"
+	"github.com/cockroachdb/cockroach/pkg/sql/sqlbase"
+	"github.com/cockroachdb/cockroach/pkg/util/ctxgroup"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+)
+
+type mysqldumpReader struct {
+	conv     rowConverter
+	debugRow func(tree.Datums)
+}
+
+var _ inputConverter = &mysqldumpReader{}
+
+func newMysqldumpReader(
+	kvCh chan kvBatch, table *sqlbase.TableDescriptor,
+) (*mysqldumpReader, error) {
+	conv, err := newRowConverter(table, kvCh)
+	if err != nil {
+		return nil, err
+	}
+	return &mysqldumpReader{conv: *conv}, err
+}
+
+func (m *mysqldumpReader) start(ctx ctxgroup.Group) {
+}
+
+func (m *mysqldumpReader) inputFinished(ctx context.Context) {
+	close(m.conv.kvCh)
+}
+
+func (m *mysqldumpReader) readFile(
+	ctx context.Context, input io.Reader, inputIdx int32, inputName string, progressFn progressFn,
+) error {
+	evalCtx := &tree.EvalContext{SessionData: &sessiondata.SessionData{Location: time.UTC}}
+
+	var count int64
+	r := bufio.NewReaderSize(input, 1024*64)
+	tokens := mysql.NewTokenizer(r)
+
+	for {
+		stmt, err := mysql.ParseNext(tokens)
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return errors.Wrap(err, "mysql parse error")
+		}
+		switch i := stmt.(type) {
+		case *mysql.Insert:
+			rows, ok := i.Rows.(mysql.Values)
+			if !ok {
+				return errors.Errorf("unexpected insert row type %T: %v", rows, i.Rows)
+			}
+			for _, inputRow := range rows {
+				count++
+				for i, raw := range inputRow {
+					converted, err := mysqlToCockroach(raw, m.conv.cols[i].DatumType(), evalCtx)
```

Comment:
can we cache the cols[i].DatumType somewhere? seems expensive to call that function for every field

Reply:
ha. nice, much better looking. thanks.